src/compiler/commandLineParser.ts

```diff
@@ -474,23 +547,31 @@ namespace ts {
             }
             else {
                 let exclude = json["exclude"] instanceof Array ? map(<string[]>json["exclude"], normalizeSlashes) : undefined;
-                let sysFiles = host.readDirectory(basePath, ".ts", exclude).concat(host.readDirectory(basePath, ".tsx", exclude));
-                for (let i = 0; i < sysFiles.length; i++) {
-                    let name = sysFiles[i];
-                    if (fileExtensionIs(name, ".d.ts")) {
-                        let baseName = name.substr(0, name.length - ".d.ts".length);
-                        if (!contains(sysFiles, baseName + ".tsx") && !contains(sysFiles, baseName + ".ts")) {
-                            fileNames.push(name);
-                        }
-                    }
-                    else if (fileExtensionIs(name, ".ts")) {
-                        if (!contains(sysFiles, name + "x")) {
-                            fileNames.push(name);
+                let extensionsToRead = getSupportedExtensions(options);
+                for (let extensionsIndex = 0; extensionsIndex < extensionsToRead.length; extensionsIndex++) {

```

Comment:
honestly logic below looks complicated.
Maybe it can be simplified to something easier to understand like:
```javascript
const extensionsByPriority = [.tsx, .ts, d.t.s] ...
let filesSeen = {};
let files = [];
for (let i = 0; i < extensionsByPriority.length; i++) {
    const ext = extensionsByPriority[i];
    let filesInDir = host.readDirectory(basePath, ext, exclude);
    // find all extensions in the list that
    // - longer than current extension
    // - ends with text of current extension
    // i.e. for ext='.ts' conflicting extension can be ['.d.ts']
    let conflictingExtensions = filter(extensionsByPriority, /*startIndex*/i + 1, e => e.length > ext.length && e.indexOf(ext, e.length - ext.length) !== -1);
    for (const fileName in filesInDir) {
        if (conflictingExtensions.some(e => fileExtensionIs(fileName, e))) {
            // file extension matches with some larger extension that was not checked yet
            // 'f.d.ts' will match '.ts' extension but really should be process later with '.d.ts' files
            continue;
        }        
        const baseName = fileName.substr(0, fileName.length - ext.length - 1);
        if (!hasProperty(filesSeen, baseName)) {
            filesSeen[baseName] = true;
            files.push(fileName);
        }
    }
}
```